pkg/compact/schedule_delete.go

```diff
@@ -0,0 +1,143 @@
+	return &ScheduleBlockDelete{
+		dir:         dir,
+		logger:      logger,
+		deleteDelay: deleteDelay,
+		bkt:         bkt,
+	}
+}
+
+// ScheduleDelete deletes blocks from bucket
+// deleteDelay duration after block is marked for deletion.
+func (s *ScheduleBlockDelete) ScheduleDelete(ctx context.Context) error {
+	compactorMetaPath := filepath.Join(s.dir, "download")
+	defer func() {
+		if err := os.RemoveAll(compactorMetaPath); err != nil {
+			level.Error(s.logger).Log("msg", "failed to remove compactor meta dir", compactorMetaPath, "err", err)
+		}
+	}()
+
+	if err := os.RemoveAll(compactorMetaPath); err != nil {
+		return errors.Wrap(err, "clean compaction compactor meta dir")
+	}
+
+	if err := objstore.DownloadDir(ctx, s.logger, s.bkt, DebugCompactorMetas, compactorMetaPath); err != nil {
+		return errors.Wrap(err, "downlad compactor-metas dir")
+	}
+
+	return filepath.Walk(compactorMetaPath, func(path string, fileInfo os.FileInfo, err error) error {
+		_, file := filepath.Split(path)
+		if _, err := ulid.Parse(strings.TrimSuffix(file, filepath.Ext(file))); err != nil {
+			compactorMetaBytes, err := ioutil.ReadFile(path)
+			if err != nil {
+				return errors.Wrap(err, "read compactor meta")
+			}
+
+			compactorMeta := CompactorMeta{}
+
+			if err := json.Unmarshal([]byte(compactorMetaBytes), &compactorMeta); err != nil {
+				return errors.Wrap(err, "unmarshal compactor meta")
+			}
+
+			if time.Now().Unix()-compactorMeta.DeletionTime > s.deleteDelay.Milliseconds() {
+				if err := block.Delete(ctx, s.logger, s.bkt, compactorMeta.ID); err != nil {
+					return errors.Wrap(err, "delete block")
+				}
+
+				if err := os.RemoveAll(filepath.Join(s.dir, compactorMeta.ID.String())); err != nil {
+					return errors.Wrap(err, "delete compactor-meta.json")
+				}
+			}
+		}
+		return nil
+	})
+}
+
+// MarkBlockForDeletion creates a file
+// which stores information about when the block was marked for deletion.
+func (s *ScheduleBlockDelete) MarkBlockForDeletion(ctx context.Context, id ulid.ULID) error {
+	compactorMetaExists, err := objstore.Exists(ctx, s.bkt, path.Join(DebugCompactorMetas, fmt.Sprintf("%s.json", id.String())))
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprint("check compactor meta for id %s in bucket", id.String()))
+	}
+	if compactorMetaExists {
+		level.Info(s.logger).Log("msg", "compactor-meta already exists for block id", id.String())
+		return nil
+	}
+
+	compactorMetaPath := filepath.Join(s.dir, "upload", fmt.Sprintf("%s.json", id.String()))
+	compactorMeta := &CompactorMeta{
+		ID:           id,
+		DeletionTime: time.Now().Unix(),
+	}
+
+	f, err := os.Create(compactorMetaPath)
+	if err != nil {
+		return err
+	}
+
+	enc := json.NewEncoder(f)
+	enc.SetIndent("", "\t")
+
+	if err := enc.Encode(compactorMeta); err != nil {
+		runutil.CloseWithLogOnErr(s.logger, f, "write meta file close")
+		return err
```

Comment:
There are some naked `return err` in this method. Would it be worth it to include the context as well?